runtime/tracing/tracing-core/common/src/aws/smithy/kotlin/runtime/tracing/CoroutineContextUtils.kt

```diff
@@ -4,224 +4,211 @@
+ * @param traceSpan The active trace span for this context.
  */
-public object TracingContext {
-    /**
-     * The active [TraceSpan] that receives events.
-     */
-    public val TraceSpan: AttributeKey<TraceSpan> = AttributeKey("TraceSpan")
-}
-
-/**
- * Gets the active [TraceSpan] for this [ExecutionContext].
- */
-public val ExecutionContext.traceSpan: TraceSpan
-    get() = get(TracingContext.TraceSpan)
+public class TraceSpanContextElement(public val traceSpan: TraceSpan) : CoroutineContext.Element {
+    override val key: CoroutineContext.Key<*>
+        get() = TraceSpanContextElement
 
-/**
- * Pushes a new child [TraceSpan] into this [ExecutionContext]. The child span will have the current span as its parent.
- * This call should be paired with a later call to [popChildTraceSpan].
- * @param The id of the child span.
- */
-public fun ExecutionContext.pushChildTraceSpan(id: String) {
-    val existingSpan = checkNotNull(getOrNull(TracingContext.TraceSpan)) { "Missing an active trace span" }
-    set(TracingContext.TraceSpan, existingSpan.child(id))
+    public companion object : CoroutineContext.Key<TraceSpanContextElement>
 }
 
 /**
- * Closes and pops the current child [TraceSpan] from this [ExecutionContext], restoring the parent span as the active
- * span. This call should be paired with a prior call to [pushChildTraceSpan].
+ * Gets the active [TraceSpan] from this [CoroutineContext].
  */
-public fun ExecutionContext.popChildTraceSpan() {
-    val existingSpan = checkNotNull(getOrNull(TracingContext.TraceSpan)) { "Missing an active trace span" }
-    existingSpan.close()
-
-    val parentSpan = checkNotNull(existingSpan.parent) { "The active trace span has no parent and cannot be popped" }
-    set(TracingContext.TraceSpan, parentSpan)
-}
+public val CoroutineContext.traceSpan: TraceSpan
+    get() = get(TraceSpanContextElement)!!.traceSpan
 
 /**
- * Runs a block of code within the context of a child [TraceSpan]. This call pushes the new child trace span into the
- * context before executing [block] and restores the parent span after the block completes (whether successfully or
- * exceptionally).
- * @param id The id of the new child span.
- * @param block The block of code to execute. Once the block is complete, the previous span will be restored.
+ * Runs a block of code within the context of a child [TraceSpan]. This call pushes the new child trace span onto the
+ * coroutine context before executing [block] and restores the parent span after the block completes (whether
+ * successfully or exceptionally).
  */
-public inline fun <R> ExecutionContext.withChildSpan(id: String, block: () -> R): R {
-    pushChildTraceSpan(id)
-    try {
-        return block()
+public suspend inline fun <R> CoroutineContext.withChildTraceSpan(
+    id: String,
+    crossinline block: suspend CoroutineScope.() -> R,
+): R {
+    val existingSpan = checkNotNull(get(TraceSpanContextElement)?.traceSpan) { "Missing an active trace span" }
```

Comment:
Should this error or should it just create a no-op trace span?

That seems to be what open telemetry does FWIW:

[Span.current()](https://javadoc.io/static/io.opentelemetry/opentelemetry-api/1.19.0/io/opentelemetry/api/trace/Span.html#current())
> Returns the [Span](https://javadoc.io/static/io.opentelemetry/opentelemetry-api/1.19.0/io/opentelemetry/api/trace/Span.html) from the current Context, falling back to a default, no-op [Span](https://javadoc.io/static/io.opentelemetry/opentelemetry-api/1.19.0/io/opentelemetry/api/trace/Span.html) if there is no span in the current context.

I think that most library code probably shouldn't care. If it's not instrumented with a parent span somewhere then it's a no-op seems fine to me. Only when the calling environment provides the span (e.g. when we generate operation setup) will it be instrumented. 

e.g.:

```kotlin
public suspend inline fun <R> span(
    id: String,
    crossinline block: suspend () -> R
): R {
    val currentSpan = // get current or no-op span
    val childSpan = currentSpan.child(id)
    return try {
        withContext(TraceSpanContext(childSpan) {
            block()
        }
     } finally {
         childSpan.close()
     }
}
```

Use:

```kotlin
suspend fun foo() = span("foo") { ... }
```